### ISSUE TITLE
cast to buffer failed

### ISSUE THREAD

**Issue by jonathanong** [Automattic/mongoose#3031]
i'm having logs that say that a cast to buffer failed:

```
May 29 14:05:40 studio app/web.2:  ValidationError: Video validation failed 
May 29 14:05:40 studio app/web.2:      at model.Document.invalidate (/app/node_modules/mongoose/lib/document.js:1156:32) 
May 29 14:05:40 studio app/web.2:      at init (/app/node_modules/mongoose/lib/document.js:317:18) 
May 29 14:05:40 studio app/web.2:      at model.Document.init (/app/node_modules/mongoose/lib/document.js:273:3) 
May 29 14:05:40 studio app/web.2:      at completeOne (/app/node_modules/mongoose/lib/query.js:1469:10) 
May 29 14:05:40 studio app/web.2:      at Immediate.<anonymous> (/app/node_modules/mongoose/lib/query.js:1204:13) 
May 29 14:05:40 studio app/web.2:      at Immediate._onImmediate (/app/node_modules/mongoose/node_modules/mquery/lib/utils.js:137:16) 
May 29 14:05:40 studio app/web.2:      at processImmediate [as _immediateCallback] (timers.js:368:17) 
May 29 14:05:40 studio app/web.2:  { sha256:  
May 29 14:05:40 studio app/web.2:     { [ValidatorError: Cast to buffer failed for value "zÔ,�e@ڤ�� 
May 29 14:05:40 studio app/web.2:     	����{=�T��u��W�/�" at path "sha256"] 
May 29 14:05:40 studio app/web.2:       properties:  
May 29 14:05:40 studio app/web.2:        { path: 'sha256', 
May 29 14:05:40 studio app/web.2:          message: 'Cast to buffer failed for value "zÔ,�e@ڤ�\u0010�\n\t��\u0002��{=�T��u��W�/�" at path "sha256"',  
May 29 14:05:40 studio app/web.2:          type: 'cast', 
May 29 14:05:40 studio app/web.2:          value: [Object] }, 
May 29 14:05:40 studio app/web.2:       message: 'Cast to buffer failed for value "zÔ,�e@ڤ�\u0010�\n\t��\u0002��{=�T��u��W�/�" at path "sha256"', 
May 29 14:05:40 studio app/web.2:       name: 'ValidatorError', 
May 29 14:05:40 studio app/web.2:       kind: 'cast', 
May 29 14:05:40 studio app/web.2:       path: 'sha256', 
May 29 14:05:40 studio app/web.2:       value:  
May 29 14:05:40 studio app/web.2:        { _bsontype: 'Binary', 
May 29 14:05:40 studio app/web.2:          sub_type: 0, 
May 29 14:05:40 studio app/web.2:          position: 32, 
May 29 14:05:40 studio app/web.2:          buffer: <Buffer 7a c3 94 2c bd 65 40 da a4 c5 10 93 0a 09 8c e9 02 db db 7b 3d f7 54 ad e7 75 b4 b1 57 f3 2f 93> } } } 
```

i'm not sure what's causing this. i have `sha256: Buffer` in my schema. seems like `sha256` is being set as a BSON Binary, which isn't exactly a buffer, causing this error. i also can't reproduce locally - it only happens on my server.

know what's going on?

**Comment by vkarpov15** [comment]
Hmm haven't seen this before. How are you generating the doc that's failing validation?

**Comment by jonathanong** [comment]
i use https://github.com/stream-utils/hash-stream to get the buffer then do:

```js
hashStream(filename, 'sha256').then(function (buf) {
  video.sha256 = buf;
  return video.save();
})
```

with the video schema something like:

```js
{
  // ...
  sha256: Buffer,
  // ...
}
```

**Comment by jonathanong** [comment]
seems like what's happening is that i'm getting the video:

```js
Video.findById(video_id).exec().then(function (video) {

})
```

but for some reason, `video.sha256` is `undefined`. then i try to save it:

```js
video.something = true;
video.save();
```

this is when the error occurs.

trying to investigate why `.sha256` is undefined. i see `"sha256" : BinData(0,"JYO5JvDJ47TMf7WrS8wzU7ELbUnzjsbdVRNDptWYwx0="),` in mongodb

**Comment by jonathanong** [comment]
downgraded to mongoose 4.0.3 and it worked.

**Comment by jonathanong** [comment]
i don't see anything that could have caused this from https://github.com/Automattic/mongoose/compare/4.0.3...4.0.4

going to downgrade to 4.0.3 for now.

**Comment by 6uliver** [comment]
I think I have the same issue. I used 3.8.0, but after upgrade to 4.0.4 my Buffer fields don't work. When I get my model these fields are undefined.

**Comment by vkarpov15** [comment]
@jonathanong @6uliver do you guys have mongodb installed as a sibling dependency? This might be related to #3020 etc.

**Comment by jonathanong** [comment]
yes. i use mongodb as well as mongoose.

**Comment by 6uliver** [comment]
Yes, me too.

**Comment by vkarpov15** [comment]
I strongly suspect this is an issue with #3020 - can you try with master? I'll see if I can repro later today

**Comment by vkarpov15** [comment]
Repro-ed, it's another instance of #3020, #2951, and #2988. Fixed in https://github.com/Automattic/mongoose/commit/1c69036088a1823fd5e131fcb3e56573ec84130f . This issue happened because we tried to do some dependency finagling to make the browser build work properly, so we bypassed the mongodb driver to get bson's instance of ObjectId and Binary so we didn't have to require() the driver in any code that did validation. Unfortunately, that meant that if you had mongodb as a sibling dependency and didn't have bson as a sibling dependency, the mongodb driver and mongoose would use different versions of bson.

You could actually work around this issue in 4.0.4 by adding `"bson": "~0.3"` to your `package.json`, but won't be necessary because this will be fixed in 4.0.5, which will be released shortly.
